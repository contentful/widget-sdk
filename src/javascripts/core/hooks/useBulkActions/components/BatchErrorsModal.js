import React from 'react';
import {
  Modal,
  Typography,
  Paragraph,
  Subheading,
  Button,
  Table,
  TableBody,
  TableCell,
  TableHead,
  TableRow,
  TextLink,
} from '@contentful/forma-36-react-components';
import PropTypes from 'prop-types';
import * as Navigator from 'states/Navigator';
import { ModalLauncher } from '@contentful/forma-36-react-components';

async function navigateToEntity(entity) {
  const entityRef = Navigator.makeEntityRef(entity);
  await Navigator.go(entityRef);
}

export function BatchErrorsModal({ successMessage, errorMessages, onClose }) {
  return (
    <>
      <Modal.Header title={'Bulk action results'} onClose={onClose} />
      <Modal.Content>
        <Typography>
          {successMessage && <Subheading>Success</Subheading>}
          <Paragraph>{successMessage}</Paragraph>
          <Subheading>Failure</Subheading>
          {errorMessages.map(([entities, errorMessage]) => (
            <div key={errorMessage}>
              <Paragraph>{errorMessage}</Paragraph>
              <Table>
                <TableHead>
                  <TableRow>
                    <TableCell>Name</TableCell>
                  </TableRow>
                </TableHead>
                <TableBody>
                  {entities.map(([entity, entityTitle]) => {
                    return (
                      <TableRow key={entity.sys.id}>
                        <TableCell>
                          <TextLink
                            onClick={(e) => {
                              e.preventDefault();
                              navigateToEntity(entity);
                              onClose();
                            }}>
                            {entityTitle}
                          </TextLink>
                        </TableCell>
                      </TableRow>
                    );
                  })}
                </TableBody>
              </Table>
            </div>
          ))}
        </Typography>
      </Modal.Content>
      <Modal.Controls>
        <Button testId="close-modal" onClick={onClose} buttonType="muted">
          Ok, close
        </Button>
      </Modal.Controls>
    </>
  );
}

export const openBatchErrorsModal = ({ successMessage, errorMessages }) => {
  return ModalLauncher.open(({ isShown, onClose }) => (
    <Modal testId="bulk-actions-fail-modal" isShown={isShown} onClose={onClose}>
      {() => (
        <BatchErrorsModal
          successMessage={successMessage}
          errorMessages={errorMessages}
          onClose={onClose}
        />
      )}
    </Modal>
  ));
};

BatchErrorsModal.propTypes = {
  successMessage: PropTypes.string,
  errorMessages: PropTypes.array,
  onClose: PropTypes.func.isRequired,
};
