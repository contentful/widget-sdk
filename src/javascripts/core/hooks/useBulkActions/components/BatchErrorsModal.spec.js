import React from 'react';
import { mount, shallow } from 'enzyme';
import { BatchErrorsModal } from './BatchErrorsModal';
import 'jest-enzyme';

describe('BatchErrorsModal', () => {
  it('should render correctly', async () => {
    const mockEntry = {
      sys: { type: 'Entry', id: 'some-entry-id' },
    };
    const wrapper = shallow(
      <BatchErrorsModal
        successMessage="1 entry saved successfully"
        errorMessages={[
          [[[mockEntry, 'Entry Title']], 'Failed to save 1 entry because a newer version exists.'],
        ]}
        isShown={true}
        onClose={jest.fn()}
      />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('call provided function when close button is clicked', async () => {
    const mockEntry = {
      sys: { type: 'Entry', id: 'some-entry-id' },
    };
    const onCloseStub = jest.fn();

    const wrapper = mount(
      <BatchErrorsModal
        successMessage="1 entry saved successfully"
        errorMessages={[
          [[[mockEntry, 'Entry Title']], 'Failed to save 1 entry because a newer version exists.'],
        ]}
        isShown={true}
        onClose={onCloseStub}
      />
    );
    const closeButton = wrapper.findWhere((el) => el.prop('testId') === 'close-modal');

    expect(closeButton.prop('onClick')).toEqual(onCloseStub);
  });
});
