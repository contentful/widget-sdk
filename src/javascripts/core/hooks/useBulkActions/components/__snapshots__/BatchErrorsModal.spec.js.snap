// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BatchErrorsModal should render correctly 1`] = `
<Fragment>
  <ModalHeader
    onClose={[MockFunction]}
    title="Bulk action results"
  />
  <ModalContent>
    <Typography>
      <Subheading>
        Success
      </Subheading>
      <Paragraph>
        1 entry saved successfully
      </Paragraph>
      <Subheading>
        Failure
      </Subheading>
      <div
        key="Failed to save 1 entry because a newer version exists."
      >
        <Paragraph>
          Failed to save 1 entry because a newer version exists.
        </Paragraph>
        <Table>
          <TableHead>
            <TableRow>
              <TableCell>
                Name
              </TableCell>
            </TableRow>
          </TableHead>
          <TableBody>
            <TableRow
              key="some-entry-id"
            >
              <TableCell>
                <TextLink
                  onClick={[Function]}
                >
                  Entry Title
                </TextLink>
              </TableCell>
            </TableRow>
          </TableBody>
        </Table>
      </div>
    </Typography>
  </ModalContent>
  <ModalControls>
    <Button
      buttonType="muted"
      onClick={[MockFunction]}
      testId="close-modal"
    >
      Ok, close
    </Button>
  </ModalControls>
</Fragment>
`;
